### ISSUE TITLE
`OSC 4` support for setting palette colors

### ISSUE THREAD

**Issue by christianparpart** [christianparpart/contour#221]
excerpt spec from the xterm ctlseqs.txt for `OSC 4`:
```
Ps = 4 ; c ; spec ⇒  Change Color Number c to the color
          specified by spec.  This can be a name or RGB specification as
          per XParseColor.  Any number of c/spec pairs may be given.
          The color numbers correspond to the ANSI colors 0-7, their
          bright versions 8-15, and if supported, the remainder of the
          88-color or 256-color table.

          If a "?" is given rather than a name or RGB specification,
          xterm replies with a control sequence of the same form which
          can be used to set the corresponding color.  Because more than
          one pair of color number and specification can be given in one
          control sequence, xterm can make more than one reply.
```

multiple colors can be set by separating with `;`.

**Comment by christianparpart** [comment]
merged.

**Comment by subnut** [comment]
Looks like it is not implemented right now -

https://github.com/contour-terminal/contour/blob/4f72517d8f5a2309e5b5267aa96e5c6b0bca4b62/src/terminal/Functions.h#L360-L361

Re-open this issue?

**Comment by uspasojevic96** [comment]
@subnut Umm, if this is bugged let us know, but this is implemented:

https://github.com/contour-terminal/contour/blob/master/src/terminal/Sequencer.cpp#L617-L632

**Comment by subnut** [comment]
@uspasojevic96 

On Ubuntu 18.04 LTS using `contour v0.2.3.183` installed from here [here](https://github.com/contour-terminal/contour/releases/download/v0.2.3.183/contour_0.2.3.183-ubuntu_18_04_amd64.deb)

```
vt.parser:Sequencer.cpp:1283] Invalid VT sequence: OSC 4 0;#eee8d5
[vt.parser:Sequencer.cpp:1283] Invalid VT sequence: OSC 4 1;#dc322f
[vt.parser:Sequencer.cpp:1283] Invalid VT sequence: OSC 4 2;#859900
[vt.parser:Sequencer.cpp:1283] Invalid VT sequence: OSC 4 3;#b58900
[vt.parser:Sequencer.cpp:1283] Invalid VT sequence: OSC 4 4;#268bd2
[vt.parser:Sequencer.cpp:1283] Invalid VT sequence: OSC 4 5;#d33682
[vt.parser:Sequencer.cpp:1283] Invalid VT sequence: OSC 4 6;#2aa198
[vt.parser:Sequencer.cpp:1283] Invalid VT sequence: OSC 4 7;#073642
[vt.parser:Sequencer.cpp:1283] Invalid VT sequence: OSC 4 8;#6c7c80
[vt.parser:Sequencer.cpp:1283] Invalid VT sequence: OSC 4 9;#dc322f
[vt.parser:Sequencer.cpp:1283] Invalid VT sequence: OSC 4 10;#859900
[vt.parser:Sequencer.cpp:1283] Invalid VT sequence: OSC 4 11;#b58900
[vt.parser:Sequencer.cpp:1283] Invalid VT sequence: OSC 4 12;#268bd2
[vt.parser:Sequencer.cpp:1283] Invalid VT sequence: OSC 4 13;#d33682
[vt.parser:Sequencer.cpp:1283] Invalid VT sequence: OSC 4 14;#2aa198
[vt.parser:Sequencer.cpp:1283] Invalid VT sequence: OSC 4 15;#073642
[vt.parser:Sequencer.cpp:1283] Invalid VT sequence: OSC 10 #002b36
[vt.parser:Sequencer.cpp:1283] Invalid VT sequence: OSC 11 #eee8d5
[vt.parser:Sequencer.cpp:1283] Invalid VT sequence: OSC 12 #dc322f
```

**Comment by christianparpart** [comment]
@subnut if there is a bug, I'm very interested in fixing it. Also. There was a notification about some comment that I just cannot see here. Did you delete it? I am curious what software you used to trigger t hat. If it's not a bug on our side, we could still escalate that to the other software's developer, such that they can fix it. :)

**Comment by subnut** [comment]
Sorry for the unnecessary noise, @uspasojevic96 and @christianparpart :sweat_smile:

The issue was that the program ([`kfc`](https://github.com/mcpcpc/kfc)) I was using to set the terminal colors was setting the colors in `#rrggbb` format, but `contour` wants them in `rgb:rr/gg/bb` format (I discovered that by running `OSC;4;0;?`). I had not noticed that difference before making the comment above.

---

@christianparpart The notification you received must have been for the _Hidden Comment_ above.

![Screenshot from 2022-01-31 22-12-50](https://user-images.githubusercontent.com/56714626/151835606-8d6a607e-7233-4cc4-80c7-bd876561e4d3.png "Hidden Comment")

**Comment by christianparpart** [comment]
@subnut this is a missing implementation indeed. There is sadly a huge design flaw in OSC 4, that is, it says it is to support whatever XParseColor (from Xlib API) supports. And that's not just one trivial format but way to many. It's just guess-work to see what is used by the client apps. Most likely what was already implemented and apparently also `#RRGGBB` by at least `kfc`. So thanks for the tip.

I've created a new PR for quickly adding that support.

I noticed you are on 0.2.3. Which is great, it's the latest stable release and I wanted to release a 0.2.4 also. But most likely I'll release 0.3.0 before and then I'm not sure how much priority a 0.2.4 should have.

The above linked PR will at least go into master branch.
